sdk/metric/sdk.go

```diff
@@ -336,39 +346,100 @@ func DefaultErrorHandler(err error) {
 // Labels returns a LabelSet corresponding to the arguments.  Passed
 // labels are de-duplicated, with last-value-wins semantics.
 func (m *SDK) Labels(kvs ...core.KeyValue) api.LabelSet {
-	// Note: This computes a canonical encoding of the labels to
-	// use as a map key.  It happens to use the encoding used by
-	// statsd for labels, allowing an optimization for statsd
-	// batchers.  This could be made configurable in the
-	// constructor, to support the same optimization for different
-	// batchers.
-
 	// Check for empty set.
 	if len(kvs) == 0 {
 		return &m.empty
 	}
 
-	ls := &labels{
-		meter:  m,
-		sorted: kvs,
+	ls := &labels{ // allocation
+		meter: m,
+		slice: kvs,
 	}
 
-	// Sort and de-duplicate.
-	sort.Stable(&ls.sorted)
+	// Sort and de-duplicate.  Note: this use of `ls.slice` avoids
+	// an allocation by using the address-able field rather than
+	// `kvs`.  Labels retains a copy of this slice, i.e., the
+	// initial allocation at the varargs call site.
+	//
+	// Note that `ls.slice` continues to refer to this memory,
+	// even though a new array is allocated for `ls.ordered`.  It
+	// is possible for the `slice` to refer to the same memory,
+	// although in the reflection code path of `computeOrdered` it
+	// costs an allocation to yield a slice through
+	// `(reflect.Value).Interface()`.
+	//
+	// TODO: It would be better overall if the export.Labels interface
+	// did not expose a slice via `Ordered()`, if instead it exposed
+	// getter methods like `Len()` and `Order(i int)`.  Then we would
+	// just implement the interface using the `orderedLabels` array.
+	sort.Stable(&ls.slice)
+
 	oi := 1
-	for i := 1; i < len(ls.sorted); i++ {
-		if ls.sorted[i-1].Key == ls.sorted[i].Key {
-			ls.sorted[oi-1] = ls.sorted[i]
+	for i := 1; i < len(kvs); i++ {
+		if kvs[i-1].Key == kvs[i].Key {
+			kvs[oi-1] = kvs[i]
 			continue
 		}
-		ls.sorted[oi] = ls.sorted[i]
+		kvs[oi] = kvs[i]
 		oi++
 	}
-	ls.sorted = ls.sorted[0:oi]
+	kvs = kvs[0:oi]
+	ls.slice = kvs
+	ls.computeOrdered(kvs)
+	return ls
+}
 
-	ls.encoded = m.labelEncoder.Encode(ls.sorted)
+func (ls *labels) computeOrdered(kvs []core.KeyValue) {
```

Comment:
This looks very nice, do we have a test to prove that this works?

Reply:
😀 I am referring to `TestSDKLabelsDeduplication`.